### PR TITLE
feat: add InMemorySpanExporter

### LINKED ISSUE #1195: In-Memory SpanExporter
Opening an issue to track adding InMemory SpanExporter, similar to whats done for [metrics](https://github.com/open-telemetry/opentelemetry-rust/issues/961).

### PR DESCRIPTION
**Pull request by zengxilong**
Fixes #1195

## Changes

* Add in memory span exporter, which can be used to inspect the `SpanData` exported using `get_finished_spans`

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by zengxilong** [comment]
The pub mod `trace` in `opentelemetry-sdk/testing` already exists，so I add a mod `span`. Is this appropriate?

File: opentelemetry-sdk/src/testing/span/in_memory_exporter.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    /// ```
+    /// # use opentelemetry_sdk::testing::span::InMemorySpanExporter;
+    ///
+    /// let exporter = InMemorySpanExporter::default();
+    /// let finished_spans = exporter.get_finished_spans().unwrap();
+    /// ```
+    pub fn get_finished_spans(&self) -> TraceResult<Vec<SpanData>> {
+        self.spans
+            .lock()
+            .map(|spans_guard| spans_guard.iter().map(Self::clone_span).collect())
```
Comment (AdamKaczmarski) [inline opentelemetry-sdk/src/testing/span/in_memory_exporter.rs:111]: Check if you can use `iter().cloned().collect()` 
ref:
[comment](https://github.com/open-telemetry/opentelemetry-rust/pull/1231#discussion_r1307589254)
[my commit](https://github.com/open-telemetry/opentelemetry-rust/pull/1231/commits/03303122f5aaeb6603689f35332aab6dd217a945)

File: opentelemetry-sdk/Cargo.toml [hunk 2]
```diff
@@ -48,7 +48,8 @@ jaeger_remote_sampler = ["trace", "opentelemetry-http", "http", "serde", "serde_
 logs = ["opentelemetry_api/logs", "crossbeam-channel", "async-trait", "serde_json"]
 logs_level_enabled = ["logs", "opentelemetry_api/logs_level_enabled"]
 metrics = ["opentelemetry_api/metrics", "regex", "async-trait"]
-testing = ["opentelemetry_api/testing", "trace", "metrics", "logs", "rt-async-std", "rt-tokio", "rt-tokio-current-thread", "tokio/macros", "tokio/rt-multi-thread"]
+span = ["opentelemetry_api/trace"]
```
Comment (lalitb) [inline opentelemetry-sdk/Cargo.toml:51]: A separate 'span' feature is unnecessary, given that we already have a 'trace' feature.

File: opentelemetry-sdk/Cargo.toml [hunk 3]
```diff
@@ -48,7 +48,8 @@ jaeger_remote_sampler = ["trace", "opentelemetry-http", "http", "serde", "serde_
...
+testing = ["opentelemetry_api/testing", "trace", "metrics","span", "logs", "rt-async-std", "rt-tokio", "rt-tokio-current-thread", "tokio/macros", "tokio/rt-multi-thread"]
```
Comment (lalitb) [inline opentelemetry-sdk/Cargo.toml:52]: `span` feature can be removed from here too.

File: opentelemetry-sdk/src/testing/span/in_memory_exporter.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    ///
+    /// let exporter = InMemorySpanExporter::default();
+    /// exporter.reset();
+    /// ```
+    pub fn reset(&self) {
+        let _ = self.spans.lock().map(|mut spans_guard| spans_guard.clear());
+    }
+
+    fn clone_span(span: &SpanData) -> SpanData {
+        span.clone()
```
Comment (lalitb) [inline opentelemetry-sdk/src/testing/span/in_memory_exporter.rs:130]: This method is perhaps not required. Refer to the earlier comment.

File: opentelemetry-sdk/src/testing/mod.rs [hunk 5]
```diff
@@ -3,3 +3,6 @@ pub mod trace;
 
 #[cfg(all(feature = "testing", feature = "metrics"))]
 pub mod metrics;
+
+#[cfg(all(feature = "testing", feature = "span"))]
+pub mod span;
```
Comment (lalitb) [inline opentelemetry-sdk/src/testing/mod.rs:8]: I believe there's no need for an independent `span` feature as mentioned earlier. The in_memory_exporter can be integrated within the `trace` feature. Create a new directory named `trace` and relocate the existing `trace.rs` and `in_memory_exporter.rs` files into it. Once that's done, we can delete the `span` directory.

**Review by cijothomas — APPROVED** [review]
LGTM!

**Comment by cijothomas** [comment]
@zengxilong Could you clear the CI checks ? They are failing now.

**Comment by zengxilong** [comment]
> @zengxilong Could you clear the CI checks ? They are failing now.

I'll fix it

File: opentelemetry-sdk/src/testing/trace/in_memory_exporter.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+/// ```
+///# use opentelemetry::trace::{SpanKind, TraceContextExt};
+///# use opentelemetry::{global, trace::Tracer, Context};
+///# use opentelemetry_sdk::propagation::TraceContextPropagator;
+///# use opentelemetry_sdk::runtime;
+///# use opentelemetry_sdk::testing::trace::InMemorySpanExporterBuilder;
+///# use opentelemetry_sdk::trace::{BatchSpanProcessor, TracerProvider};
+///
+///# #[tokio::main]
+///# async fn main() {
```
Comment (shaun-cox) [inline opentelemetry-sdk/src/testing/trace/in_memory_exporter.rs:21]: Is the line below needed?  It doesn't appear this test uses context propagation...

File: opentelemetry-sdk/src/testing/trace/in_memory_exporter.rs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+///     global::set_text_map_propagator(TraceContextPropagator::new());
```
Comment (shaun-cox) [inline opentelemetry-sdk/src/testing/trace/in_memory_exporter.rs:22]: Is the line above needed? It doesn't appear this test uses context propagation...

Comment (zengxilong) [reply opentelemetry-sdk/src/testing/trace/in_memory_exporter.rs:22]: You're right，I‘ll delete this line.

**Review by shaun-cox — APPROVED** [review]

**Pull request merged**